[aksel.nav.no] Add possibility to import code in examples

HalvorHaugan opened the pull request:
It's very cumbersome to copy-paste code across templates, so these changes makes it possible to import code into templates/examples.

Some of the code is a bit complicated, especially the code related to jscodeshift, so I have tried to add comments to make it easier to read. Please let me know if you have any ideas to improve readability (or just push the changes yourself).

It works by dynamically injecting the imported code when pushing the examples to Sanity. Read comments in code for details.

Todo later: Refactor all templates (only did a few here to reduce the amount of changes to review).

HalvorHaugan committed:
[aksel.nav.no] Add possibility to import code in examples

HalvorHaugan committed:
fix comments

HalvorHaugan committed:
fix filter-code.test.ts

HalvorHaugan committed:
Merge branch 'main' into template-magic

KenAJoh committed:
feat: Added custom eslint rule for imports in templates

HalvorHaugan committed:
more eslint rules

HalvorHaugan committed:
Moved __parts*

HalvorHaugan committed:
Forgot this when moving parts

KenAJoh commented on .eslintrc.js:
```diff
@@ -33,6 +33,17 @@ module.exports = {
+          },
+        ],
+      },
+    ],
```
```suggestion
```

KenAJoh commented on aksel.nav.no/website/scripts/update-examples/parts/filter-code.ts:
```diff
@@ -1,23 +1,164 @@
+  const jsxToReplace: Record<string, string> = {};
+  const codeToInject: string[] = [];
+
+  importDeclarations.forEach((importDecl) => {
```
```suggestion
  importDeclarations.forEach((importDecl: : ASTPath<ImportDeclaration>) => {
```

KenAJoh replied:
import jscodeshift, { ASTPath, Collection, ImportDeclaration } from "jscodeshift";

HalvorHaugan replied:
Y tho?

KenAJoh replied:
is infered as `any` for me, does TS correctly infer types here for you?

KenAJoh replied:
Nvm seems like an issue on my side, after a clean install + restart ts server it infered correctly. Can ignore these changes then 👍

KenAJoh commented on aksel.nav.no/website/scripts/update-examples/parts/filter-code.ts:
```diff
@@ -1,23 +1,164 @@
+    const parsedImportedCode = j(importedCode);
+
+    // Imports from the __parts-inline directory are components that should be inlined in the example JSX.
+    if (importPath.includes("/__parts-inline/")) {
```
For first time readers of this code, it might be easier if the values
```
__parts-inline,
__parts,
react 
@navikt/ds-react
```
that are used as constant are placed above the function with a small comment on what it's used for

```js
/* Comment for what these two do */
const inlineImport = "__parts-inline",
const replaceImport = "__parts"; 

/* To allow easier copy-paste of templates and examples, we want to make sure no extra 3rd-party packages are needed */
const allowedPackages = ["react"];
const allowedPackageScope = ["@navikt"];
```

HalvorHaugan replied:
Done. I hope the comments makes sense.

KenAJoh commented on aksel.nav.no/website/scripts/update-examples/parts/filter-code.ts:
```diff
@@ -1,23 +1,164 @@
+  parsedImportedCode.find(j.ImportDeclaration).forEach((nestedImportDecl) => {
+    const packageName = nestedImportDecl.value.source.value?.toString() || "";
+    if (
+      (!packageName.includes("@navikt/") && packageName !== "react") ||
```
If i'm not mistaken this code is also ran for all the examples? If so we need to make a quick check and fix some of the stories. Some of the datepicker/monthopicker uses `date-fns` as an import last time i checked.

HalvorHaugan replied:
This particular code only runs on files in __parts and __parts-inline.

HalvorHaugan committed:
Update .eslintrc.js

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

HalvorHaugan committed:
fix eslint

KenAJoh committed:
bug: using test.describe broke async-tests with playwright. Now just uses test()

KenAJoh committed:
bug: Playwright did not allow dynamic generation of testnames when they were async

Fix: Tests are now fetched and generated with sync, avoiding any async functions before the test itself runs

HalvorHaugan commented on aksel.nav.no/website/e2e/smoketest-examples.test.ts:
Considering renaming this file to "smoketest-sandbox.e2e.ts", to correspond with "sandbox.e2e.ts". What do you think?

KenAJoh replied:
Go for it 🚀

HalvorHaugan committed:
client.server.js -> ts

HalvorHaugan committed:
Simplify error message in import-check

HalvorHaugan committed:
Attempt to fix Chromatic

HalvorHaugan committed:
Attempt 2 to fix Chromatic

HalvorHaugan committed:
Attempt to fix playwright

HalvorHaugan committed:
Attempt 3 to fix Chromatic

HalvorHaugan committed:
Attempt 2 to fix PW

HalvorHaugan committed:
Merge branch 'main' into template-magic

HalvorHaugan committed:
Rename smoketest-examples.test.ts

HalvorHaugan committed:
Fix error when merging imports

HalvorHaugan committed:
gitignore ___temp

HalvorHaugan committed:
Validate code

HalvorHaugan committed:
better comments (hopefully)

KenAJoh commented on .eslintrc.js:
```diff
@@ -91,6 +91,31 @@ module.exports = {
+    },
+    {
+      files: [
+        "aksel.nav.no/website/components/website-modules/examples/__parts-inline/*.tsx",
```
```suggestion
        "**/examples/__parts*/*.tsx",
      ],
      plugins: ["aksel-local"],
      rules: {
        "aksel-local/import-check": ["error"], // Only allow imports from @navikt and react
      },
    },
    {
      files: [
        "**/examples/__parts-inline/*.tsx",
```
Would this suffice when scoping the ruleset?

KenAJoh commented on .github/workflows/ci.yml:
```diff
@@ -37,5 +37,8 @@ jobs:
         run: yarn lint
 
+      - name: Validate example code
+        run: yarn workspace website validate:eksempler
```
```suggestion
        run: yarn validate:eksempler
```
When a script is scoped with`:`, yarn is smart enough to find it from root as long as its unique to the workspace.

KenAJoh commented on aksel.nav.no/website/scripts/update-examples/validate.ts:
```diff
@@ -0,0 +1,29 @@
+  for (const directory of rootDirectories) {
+    if (!(await validate(directory))) {
+      error = true;
+    }
```
```suggestion
    } else {
        console.log("Everything Is A-OK\n")
    }
```

HalvorHaugan committed:
Merge branch 'main' into template-magic

HalvorHaugan committed:
Update .github/workflows/ci.yml

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

HalvorHaugan committed:
Update aksel.nav.no/website/scripts/update-examples/validate.ts

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

HalvorHaugan committed:
Update .eslintrc.js

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

HalvorHaugan committed:
prettier

HalvorHaugan committed:
Adjust logging

HalvorHaugan committed:
Merge branch 'main' into template-magic

HalvorHaugan committed:
Remove redundant tests
